### PR TITLE
Add fallback-mode option for subset

### PR DESCRIPTION
**Pull request by takanabe**
# Background

When the subset API errors or returns isBrainless=true, users can now opt into 'stop' (exit non-zero to halt CI) or 'random-sample' (split locally using --target) instead of the default run-all behavior. The option is hidden and off by default so existing pipelines are unaffected.

# Changes

- Add `--fallback-mode` option for subset command

### CODE REVIEW

File: smart_tests/commands/subset.py [hunk 1]
```diff
@@ -562,6 +586,23 @@ def _collect_potential_test_files(self):
...
+        if self.fallback_mode == FallbackMode.STOP:
+            click.echo(
+                "Warning: Smart Tests could not retrieve a subset. Stopping build (--fallback-mode=stop).",
+                err=True,
+            )
+            sys.exit(1)
+        elif self.fallback_mode == FallbackMode.RANDOM_SAMPLE:
+            target_fraction = float(self.target) if self.target is not None else 1.0
+            click.echo(
+                "Warning: Smart Tests could not retrieve a subset. Falling back to local random sample.",
```
Comment (Konboi) [inline smart_tests/commands/subset.py]: Since users may use options other than --target e.g.) --confidence, it might be helpful to include the target percentage in the warning message as well.

Or, if --fallback_mode is specified and the user is using an option other than --target, we could return an error in the validation arera

File: smart_tests/commands/subset.py [hunk 2]
```diff
@@ -82,6 +89,14 @@ def from_test_paths(cls, test_paths: List[TestPath]) -> 'SubsetResult':
             is_observation=False
         )
 
+    @classmethod
+    def from_random_sample(cls, test_paths: List[TestPath], target: float) -> 'SubsetResult':
+        count = max(1, round(len(test_paths) * target))
+        sampled = random.sample(test_paths, min(count, len(test_paths)))
+        sampled_set = {id(t): t for t in sampled}
+        rest = [t for t in test_paths if id(t) not in sampled_set]
+        return cls(subset=sampled, rest=rest, subset_id='', summary={}, is_brainless=False, is_observation=False)
```
Comment (Konboi) [inline smart_tests/commands/subset.py:98]: Is `is_brainless=False` intentionally?

File: tests/commands/test_api_error.py [hunk 3]
```diff
@@ -420,3 +420,104 @@ def assert_tracking_count(self, tracking, count: int):
...
+        responses.replace(
+            responses.POST,
+            f"{get_base_url()}/intake/organizations/{self.organization}/workspaces/{self.workspace}/subset",
+            status=500)
+
+        with tempfile.NamedTemporaryFile(delete=False) as rest_file:
+            result = self.cli(*self._subset_args(rest_file.name, ("--fallback-mode", "random-sample")), mix_stderr=False)
+            self.assert_success(result)
+            all_tests = result.stdout.strip().split("\n") + Path(rest_file.name).read_text().strip().split("\n")
+            all_tests = [t for t in all_tests if t]
```
Comment (Konboi) [inline tests/commands/test_api_error.py]: Q: What is this loop?

File: smart_tests/commands/subset.py [hunk 4]
```diff
@@ -697,6 +738,11 @@ def run(self):
         # TODO(Konboi): split subset isn't provided for smart-tests initial release
         # if split:
         #   click.echo("subset/{}".format(subset_result.subset_id))
+        if subset_result.is_brainless:
```
Comment (Konboi) [inline smart_tests/commands/subset.py:740]: In Brainless mode, we're already splitting requests between subset and rest on the server side.

I can understand the stop behavior, but for sampling, do you mean re-sampling the tests on the CLI side?

File: smart_tests/commands/subset.py (hunk 2 above)
Comment (takanabe) [reply smart_tests/commands/subset.py:98]: Yes, `is_brainless=False` is intentional. The random sample result is locally generated, not from the brainless model. I hope the behavior matches with this implementation https://github.com/cloudbees-oss/smart-tests-cli/pull/1308/changes/ea8477af18cc53f219f96cc40e6bfa829ba3d279

File: smart_tests/commands/subset.py (hunk 1 above)
Comment (takanabe) [reply smart_tests/commands/subset.py]: sounds prudent. Let me consider

File: tests/commands/test_api_error.py (hunk 3 above)
Comment (takanabe) [reply tests/commands/test_api_error.py]: This is a filter. "".split("\n") returns [""] rather than []. So if either stdout or the rest file is empty, the concatenation produces empty strings and the count assertion would be wrong.

Probably, there is a better way. I'll consider again.

File: smart_tests/commands/subset.py (hunk 4 above)
Comment (takanabe) [reply smart_tests/commands/subset.py:740]: yeah, you are right. this is redundant if we split tests on server.

Comment (takanabe) [reply smart_tests/commands/subset.py:740]: handled with https://github.com/cloudbees-oss/smart-tests-cli/pull/1308/changes/ea8477af18cc53f219f96cc40e6bfa829ba3d279

File: tests/commands/test_api_error.py (hunk 3 above)
Comment (takanabe) [reply tests/commands/test_api_error.py]: I treated the empty string as the edge case but let me drop the filter for the test for simplicity https://github.com/cloudbees-oss/smart-tests-cli/pull/1308/changes/0451d3ac77fcf6b8ef9258fdadf51f76605036f6

File: smart_tests/commands/subset.py (hunk 1 above)
Comment (takanabe) [reply smart_tests/commands/subset.py]: handled with https://github.com/cloudbees-oss/smart-tests-cli/pull/1308

File: smart_tests/commands/subset.py (hunk 2 above)
Comment (Konboi) [reply smart_tests/commands/subset.py:98]: OK 👍 
~Could you add a comment that why we update the brainless mode value from true to false here?~

I misunderstood the data flow, you don't have to add comment, sorry

**Review by Konboi — APPROVED** [review]
LGTM 

---

I left a comment but it's nits comment so I'll approve this PR
